src/NHibernate/Impl/ExpressionProcessor.cs

```diff
@@ -248,32 +248,48 @@ private static ICriterion Le(ProjectionInfo property, object value)
 		}
 
 		/// <summary>
-		/// Invoke the expression to extract its runtime value
+		/// Walk or Invoke expression to extract its runtime value
 		/// </summary>
 		public static object FindValue(Expression expression)
 		{
-			if (expression.NodeType == ExpressionType.Constant)
-				return ((ConstantExpression) expression).Value;
-
-			if (expression.NodeType == ExpressionType.MemberAccess)
+			switch (expression.NodeType)
 			{
-				var memberAccess = (MemberExpression) expression;
-				if (memberAccess.Expression == null || memberAccess.Expression.NodeType == ExpressionType.Constant)
-				{
-					var constantValue = ((ConstantExpression) memberAccess.Expression)?.Value;
-					var member = memberAccess.Member;
-					switch (member.MemberType)
+				case ExpressionType.Constant:
+					var constantExpression = (ConstantExpression) expression;
+					return constantExpression.Value;
+				case ExpressionType.MemberAccess:
+					var memberExpression = (MemberExpression) expression;
+					object getObj() => memberExpression.Expression != null ? FindValue(memberExpression.Expression) : null;
+
+					switch (memberExpression.Member.MemberType)
 					{
 						case MemberTypes.Field:
-							return ((FieldInfo) member).GetValue(constantValue);
+							return ((FieldInfo) memberExpression.Member).GetValue(getObj());
 						case MemberTypes.Property:
-							return ((PropertyInfo) member).GetValue(constantValue);
+							return ((PropertyInfo) memberExpression.Member).GetValue(getObj());
 					}
-				}
+					break;
+				case ExpressionType.Call:
+					var methodCallExpression = (MethodCallExpression) expression;
+					var args = methodCallExpression.Arguments.ToArray(arg => FindValue(arg));
+
+					var callingObject = methodCallExpression.Object == null ? null : FindValue(methodCallExpression.Object);
+					return methodCallExpression.Method.Invoke(callingObject, args);
+				case ExpressionType.Convert:
+					var unaryExpression = (UnaryExpression) expression;
+
+					if (unaryExpression.Method != null)
+						return unaryExpression.Method.Invoke(null, new[] { FindValue(unaryExpression.Operand) });
+
+					if (unaryExpression.Type == typeof(object)
```

Comment:
@bahusoid 
int? value = 1234;
Expression<Func<int>> xxx = ()=> (int) value; with removing (Nullable.GetUnderlyingType(unaryExpression.Operand.Type) ?? unaryExpression.Operand.Type)  == unaryExpression.Type causes Compile.
Please look at test case GivenNullableIntegerToIntegerCast that does not checks is compiled or not bu if you debug old code that does not compile expression

Reply:
`GivenNullableIntegerToIntegerCast`  is not compiled for me with latest code.

Reply:
But I see inverted case would be compiled (int -> int?)

Reply:
Fixed. Thanks for noticing.